File: mobile/customer-app/app/src/main/java/com/sukashawarma/customer/ui/screens/catalog/KatalogFilter.kt
package com.sukashawarma.customer.ui.screens.catalog

import com.sukashawarma.customer.data.api.MenuItemDto

/** Satu kelompok menu yang tampil dengan judulnya sendiri di katalog. */
data class KategoriMenu(
    val id: String?,
    val nama: String,
    val items: List<MenuItemDto>
)

private const val NAMA_LAINNYA = "Lainnya"
private const val NAMA_KATEGORI_TANPA_NAMA = "Menu"

/**
 * Mengelompokkan katalog per kategori untuk ditampilkan.
 *
 * Tiga aturan yang tidak boleh dilanggar:
 *
 * 1. **Tidak ada item yang boleh hilang.** Item tanpa kategori, kategori tanpa
 *    nama, dan item yang sedang habis semuanya tetap terbit. Menu yang lenyap
 *    membuat pelanggan mengira menunya memang tidak pernah ada; menu yang
 *    diredupkan menjelaskan dirinya sendiri.
 * 2. **Urutan harus stabil.** Data yang sama harus menghasilkan susunan yang
 *    sama setiap kali, termasuk ketika `sort_order` seluruhnya null. Karena
 *    itu urutan kemunculan di daftar asal dipakai sebagai pemutus terakhir,
 *    bukan urutan HashMap yang tak dijamin.
 * 3. **"Lainnya" selalu paling bawah**, berapa pun angka urutan kategori lain.
 */
fun kelompokkanPerKategori(items: List<MenuItemDto>): List<KategoriMenu> {
    val kelompok = LinkedHashMap<String?, MutableList<MenuItemDto>>()
    for (it in items) {
        kelompok.getOrPut(it.categoryId) { mutableListOf() }.add(it)
    }

    val urutanMuncul = kelompok.keys.toList()

    return kelompok.entries
        .map { (id, isi) ->
            KategoriMenu(
                id = id,
                nama = when {
                    id == null -> NAMA_LAINNYA
                    else -> isi.firstNotNullOfOrNull { it.categoryName }
                        ?: NAMA_KATEGORI_TANPA_NAMA
                },
                items = isi.sortedWith(
                    // `sort_order` null berarti "belum diatur", bukan nol.
                    // Mengurutkannya sebagai nol akan melempar item yang belum
                    // diatur ke paling atas -- persis kebalikan dari maksudnya.
                    compareBy<MenuItemDto> { it.sortOrder ?: Int.MAX_VALUE }
                        .thenBy { it.name.lowercase() }
                )
            )
        }
        .sortedWith(
            compareBy<KategoriMenu> { if (it.id == null) 1 else 0 }
                .thenBy { kelompok[it.id]?.minOfOrNull { m -> m.categorySortOrder ?: Int.MAX_VALUE } ?: Int.MAX_VALUE }
                .thenBy { kelompok[it.id]?.minOfOrNull { m -> m.sortOrder ?: Int.MAX_VALUE } ?: Int.MAX_VALUE }
                .thenBy { urutanMuncul.indexOf(it.id) }
        )
}

/**
 * Menyaring katalog dengan kata kunci pencarian.
 *
 * Kueri kosong mengembalikan seluruh daftar -- bukan daftar kosong. Kekeliruan
 * arah ini membuat katalog tampak kosong begitu kolom pencarian dibersihkan.
 */
fun saringPencarian(items: List<MenuItemDto>, kueri: String): List<MenuItemDto> {
    val bersih = kueri.trim().lowercase()
    if (bersih.isEmpty()) return items
    return items.filter {
        it.name.lowercase().contains(bersih) ||
            (it.description?.lowercase()?.contains(bersih) == true)
    }
}

File: mobile/customer-ios-app/SukaShawarma/UI/Menu/KatalogFilter.swift
import Foundation

/// Satu kelompok menu yang tampil dengan judulnya sendiri di katalog.
struct KategoriMenu: Equatable, Identifiable {
    let id: String?
    let nama: String
    let items: [MenuItemDto]

    var kunci: String { id ?? "tanpa-kategori:\(nama)" }
}

private let namaLainnya = "Lainnya"
private let namaKategoriTanpaNama = "Menu"

/// Padanan `kelompokkanPerKategori` (KatalogFilter.kt). Tiga aturan:
/// 1. **Tidak ada item yang hilang** (tanpa kategori, kategori tanpa nama, habis).
/// 2. **Urutan stabil** — urutan kemunculan jadi pemutus terakhir.
/// 3. **"Lainnya" selalu paling bawah.**
func kelompokkanPerKategori(_ items: [MenuItemDto]) -> [KategoriMenu] {
    var urutanMuncul: [String?] = []
    var kelompok: [String?: [MenuItemDto]] = [:]
    for m in items {
        if kelompok[m.categoryId] == nil { urutanMuncul.append(m.categoryId) }
        kelompok[m.categoryId, default: []].append(m)
    }

    // `sort_order` nil berarti "belum diatur", bukan nol — diletakkan di akhir.
    func urut(_ n: Int?) -> Int { n ?? Int.max }

    let hasil = urutanMuncul.enumerated().map { posisi, id -> (KategoriMenu, [Int]) in
        let isi = kelompok[id] ?? []
        let nama = id == nil ? namaLainnya : (isi.lazy.compactMap(\.categoryName).first ?? namaKategoriTanpaNama)
        let terurut = isi.sorted {
            (urut($0.sortOrder), $0.name.lowercased()) < (urut($1.sortOrder), $1.name.lowercased())
        }
        let kunciUrut = [
            id == nil ? 1 : 0,
            isi.map { urut($0.categorySortOrder) }.min() ?? Int.max,
            isi.map { urut($0.sortOrder) }.min() ?? Int.max,
            posisi,
        ]
        return (KategoriMenu(id: id, nama: nama, items: terurut), kunciUrut)
    }

    return hasil.sorted { $0.1.lexicographicallyPrecedes($1.1) }.map(\.0)
}

/// Kueri kosong mengembalikan seluruh daftar — BUKAN daftar kosong.
func saringPencarian(_ items: [MenuItemDto], kueri: String) -> [MenuItemDto] {
    let bersih = kueri.trimmingCharacters(in: .whitespaces).lowercased()
    guard !bersih.isEmpty else { return items }
    return items.filter {
        $0.name.lowercased().contains(bersih) || ($0.description?.lowercased().contains(bersih) ?? false)
    }
}

/// Topping yang ditawarkan di detail menu (logika dari `AppNavigation.kt`):
/// menu berkategori "Topping" yang tersedia — kecuali untuk menu yang dirinya
/// topping atau minuman ("Suka Drink").
func toppingUntuk(_ item: MenuItemDto, semua: [MenuItemDto]) -> [MenuItemDto] {
    func kategori(_ m: MenuItemDto, _ nama: String) -> Bool {
        m.categoryName?.caseInsensitiveCompare(nama) == .orderedSame
    }
    if kategori(item, "Topping") || kategori(item, "Suka Drink") { return [] }
    return semua.filter { kategori($0, "Topping") && $0.isAvailable }
}
